src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -2,45 +2,149 @@
+               "System.Runtime",
+               "System.Runtime.CompilerServices.Unsafe",
+               "System.Runtime.Extensions",
+               "System.Runtime.InteropServices",
+               "System.Runtime.Loader",
+               "System.Runtime.Numerics",
+               "System.Runtime.Serialization.Primitives",
+               "System.Security.Cryptography.Algorithms",
+               "System.Security.Cryptography.Primitives",
+               "System.Text.Encoding.CodePages",
+               "System.Text.Encoding.Extensions",
+               "System.Text.RegularExpressions",
+               "System.Threading",
+               "System.Threading.Tasks",
+               "System.Threading.Tasks.Parallel",
+               "System.Threading.Thread",
+               "System.Threading.ThreadPool",
+               "System.Xml.ReaderWriter",
+               "System.Xml.XDocument",
+               "System.Xml.XPath.XDocument");
+
+        private readonly object _guard = new object();
+        private readonly Dictionary<string, DirectoryLoadContext> _loadContextByDirectory = new Dictionary<string, DirectoryLoadContext>(StringComparer.Ordinal);
 
         protected override Assembly LoadFromPathImpl(string fullPath)
         {
-            //.NET Native doesn't support AssemblyLoadContext.GetLoadContext. 
-            // Initializing the _loadContext in the .ctor would cause
-            // .NET Native builds to fail because the .ctor is called. 
-            // However, LoadFromPathImpl is never called in .NET Native, so 
-            // we do a lazy initialization here to make .NET Native builds happy.
-            if (_loadContext == null)
-            {
-                AssemblyLoadContext loadContext = AssemblyLoadContext.GetLoadContext(typeof(DefaultAnalyzerAssemblyLoader).GetTypeInfo().Assembly);
+            DirectoryLoadContext? loadContext;
 
-                if (System.Threading.Interlocked.CompareExchange(ref _loadContext, loadContext, null) == null)
+            var fullDirectoryPath = Path.GetDirectoryName(fullPath) ?? throw new ArgumentException(message: null, paramName: nameof(fullPath));
+            lock (_guard)
+            {
+                if (!_loadContextByDirectory.TryGetValue(fullDirectoryPath, out loadContext))
                 {
-                    _loadContext.Resolving += (context, name) =>
-                    {
-                        Debug.Assert(ReferenceEquals(context, _loadContext));
-                        return Load(name.FullName);
-                    };
+                    loadContext = new DirectoryLoadContext(fullDirectoryPath, this);
+                    _loadContextByDirectory[fullDirectoryPath] = loadContext;
                 }
             }
 
-            return LoadImpl(fullPath);
+            var name = AssemblyName.GetAssemblyName(fullPath);
+            return loadContext.LoadFromAssemblyName(name);
+        }
+
+        internal static class TestAccessor
```

Comment:
Why did you opt to have a `TestAccessor` here vs. just adding an `internal` method on `DefaultAnalyzerAssemblyLoader`?

Reply:
I wanted people to feel bad about ever accessing this in the implementation. I was attempting to imitate `TestAccessor` patterns in `ChangedText.cs`, `SegmentedArray`1.cs`, etc.